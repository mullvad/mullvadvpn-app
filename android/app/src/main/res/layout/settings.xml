<FrameLayout xmlns:android="http://schemas.android.com/apk/res/android"
             xmlns:mullvad="http://schemas.android.com/apk/res-auto"
             android:layout_width="match_parent"
             android:layout_height="match_parent"
             android:background="@color/darkBlue"
             android:gravity="left">
    <TextView android:id="@+id/title"
              android:layout_width="wrap_content"
              android:layout_height="wrap_content"
              android:text="@string/settings"
              style="@style/SettingsCollapsedHeader" />
    <LinearLayout android:layout_width="match_parent"
                  android:layout_height="match_parent"
                  android:orientation="vertical">
        <FrameLayout android:layout_width="match_parent"
                     android:layout_height="wrap_content">
            <ImageButton android:id="@+id/close"
                         android:layout_width="wrap_content"
                         android:layout_height="wrap_content"
                         android:padding="12dp"
                         android:background="?android:attr/selectableItemBackground"
                         android:src="@drawable/icon_close" />
            <TextView android:id="@+id/collapsed_title"
                      android:layout_width="wrap_content"
                      android:layout_height="wrap_content"
                      android:layout_marginHorizontal="4dp"
                      android:layout_gravity="center"
                      android:text="@string/settings"
                      style="@style/SettingsCollapsedHeader" />
        </FrameLayout>
        <net.mullvad.mullvadvpn.ui.widget.ListenableScrollView android:id="@+id/scroll_area"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="match_parent">

            <LinearLayout android:layout_width="match_parent"
                          android:layout_height="wrap_content"
                          android:orientation="vertical">
                <TextView android:id="@+id/expanded_title"
                          android:layout_width="wrap_content"
                          android:layout_height="wrap_content"
                          android:layout_marginTop="4dp"
                          android:layout_marginLeft="@dimen/side_margin"
                          android:lines="1"
                          android:text="@string/settings"
                          style="@style/SettingsExpandedHeader" />
                <net.mullvad.mullvadvpn.ui.widget.AccountCell android:id="@+id/account"
                                                              android:layout_width="match_parent"
                                                              android:layout_height="wrap_content"
                                                              android:layout_marginTop="@dimen/vertical_space"
                                                              mullvad:text="@string/settings_account" />
                <net.mullvad.mullvadvpn.ui.widget.NavigateCell android:id="@+id/preferences"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="wrap_content"
                                                               android:layout_marginTop="1dp"
                                                               mullvad:text="@string/settings_preferences" />
                <net.mullvad.mullvadvpn.ui.widget.NavigateCell android:id="@+id/advanced"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="wrap_content"
                                                               android:layout_marginTop="1dp"
                                                               mullvad:text="@string/settings_advanced" />
                <net.mullvad.mullvadvpn.ui.widget.AppVersionCell android:id="@+id/app_version"
                                                                 android:layout_width="match_parent"
                                                                 android:layout_height="wrap_content"
                                                                 android:layout_marginTop="@dimen/vertical_space"
                                                                 mullvad:text="@string/app_version"
                                                                 mullvad:footer="@string/update_available_footer" />
                <net.mullvad.mullvadvpn.ui.widget.NavigateCell android:id="@+id/report_a_problem"
                                                               android:layout_width="match_parent"
                                                               android:layout_height="wrap_content"
                                                               android:layout_marginTop="@dimen/vertical_space"
                                                               mullvad:text="@string/report_a_problem" />
                <net.mullvad.mullvadvpn.ui.widget.UrlCell android:id="@+id/faqs_and_guides"
                                                          android:layout_width="match_parent"
                                                          android:layout_height="wrap_content"
                                                          android:layout_marginTop="1dp"
                                                          mullvad:text="@string/faqs_and_guides"
                                                          mullvad:url="@string/faqs_and_guides_url" />
                <net.mullvad.mullvadvpn.ui.widget.UrlCell android:layout_width="match_parent"
                                                          android:layout_height="wrap_content"
                                                          android:layout_marginTop="1dp"
                                                          mullvad:text="@string/privacy_policy"
                                                          mullvad:url="@string/privacy_policy_url" />
            </LinearLayout>
        </net.mullvad.mullvadvpn.ui.widget.ListenableScrollView>
    </LinearLayout>
</FrameLayout>
