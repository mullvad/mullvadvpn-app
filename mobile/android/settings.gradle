rootProject.name = 'Mullvad'

include ':app', ':wireguardbinding'

