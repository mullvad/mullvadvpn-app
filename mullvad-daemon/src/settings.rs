#[cfg(not(target_os = "android"))]
use futures::TryFutureExt;
use mullvad_types::{
    relay_constraints::{RelayConstraints, RelaySettings, WireguardConstraints},
    settings::{DnsState, Settings},
};
use rand::Rng;
use std::{
    fmt::{self, Display},
    ops::Deref,
    path::{Path, PathBuf},
};
use talpid_core::firewall::is_local_address;
use talpid_types::ErrorExt;
use tokio::{
    fs,
    io::{self, AsyncWriteExt},
};

const SETTINGS_FILE: &str = "settings.json";

#[derive(err_derive::Error, Debug)]
#[error(no_from)]
pub enum Error {
    #[error(display = "Unable to read settings file {}", _0)]
    ReadError(String, #[error(source)] io::Error),

    #[error(display = "Unable to parse settings file")]
    ParseError(#[error(source)] serde_json::Error),

    #[error(display = "Unable to remove settings file {}", _0)]
    #[cfg(not(target_os = "android"))]
    DeleteError(String, #[error(source)] io::Error),

    #[error(display = "Unable to serialize settings to JSON")]
    SerializeError(#[error(source)] serde_json::Error),

    #[error(display = "Unable to write settings to {}", _0)]
    WriteError(String, #[error(source)] io::Error),
}

#[derive(Debug)]
pub struct SettingsPersister {
    settings: Settings,
    path: PathBuf,
}

pub type MadeChanges = bool;

impl SettingsPersister {
    /// Loads user settings from file. If it fails, it returns the defaults.
    pub async fn load(settings_dir: &Path) -> Self {
        let path = settings_dir.join(SETTINGS_FILE);
        let (mut settings, mut should_save) = match Self::load_from_file(&path).await {
            Ok(value) => value,
            Err(error) => {
                log::warn!(
                    "{}",
                    error.display_chain_with_msg("Failed to load settings. Using defaults.")
                );
                let mut settings = Self::default_settings();

                // Protect the user by blocking the internet by default. Previous settings may
                // not have caused the daemon to enter the non-blocking disconnected state.
                settings.block_when_disconnected = true;

                (settings, true)
            }
        };

        // If the settings file did not contain a wg_migration_rand_num then it will be initialized
        // to -1.0 by serde. This block ensures that this value is correctly intitialzed to a
        // percentage.
        if settings.wg_migration_rand_num < 0.0 || settings.wg_migration_rand_num > 1.0 {
            let mut rng = rand::thread_rng();
            settings.wg_migration_rand_num = rng.gen_range(0.0..=1.0);
            should_save |= true
        }

        // Force IPv6 to be enabled on Android
        if cfg!(target_os = "android") {
            should_save |= !settings.tunnel_options.generic.enable_ipv6;
            settings.tunnel_options.generic.enable_ipv6 = true;
        }
        if crate::version::is_beta_version() {
            should_save |= !settings.show_beta_releases;
            settings.show_beta_releases = true;
        }

        let mut persister = SettingsPersister { settings, path };

        if should_save {
            if let Err(error) = persister.save().await {
                log::error!(
                    "{}",
                    error.display_chain_with_msg("Failed to save updated settings")
                );
            }
        }

        persister
    }

    async fn load_from_file(path: &Path) -> Result<(Settings, bool), Error> {
        log::info!("Loading settings from {}", path.display());

        let settings_bytes = match fs::read(path).await {
            Ok(bytes) => bytes,
            Err(error) => {
                if error.kind() == io::ErrorKind::NotFound {
                    log::info!("No settings were found. Using defaults.");
                    return Ok((Self::default_settings(), true));
                } else {
                    return Err(Error::ReadError(path.display().to_string(), error));
                }
            }
        };
        Ok((Self::load_from_bytes(&settings_bytes)?, false))
    }

    fn load_from_bytes(bytes: &[u8]) -> Result<Settings, Error> {
        serde_json::from_slice(bytes).map_err(Error::ParseError)
    }

    async fn save(&mut self) -> Result<(), Error> {
        Self::save_inner(&self.path, &self.settings).await
    }

    /// Serializes the settings and saves them to the given file.
    async fn save_inner(path: &Path, settings: &Settings) -> Result<(), Error> {
        log::debug!("Writing settings to {}", path.display());

        let buffer = serde_json::to_string_pretty(settings).map_err(Error::SerializeError)?;
        let mut file = mullvad_fs::AtomicFile::new(path)
            .await
            .map_err(|e| Error::WriteError(path.display().to_string(), e))?;
        file.write_all(&buffer.into_bytes())
            .await
            .map_err(|e| Error::WriteError(path.display().to_string(), e))?;
        file.finalize()
            .await
            .map_err(|e| Error::WriteError(path.display().to_string(), e))?;

        Ok(())
    }

    /// Resets default settings
    #[cfg(not(target_os = "android"))]
    pub async fn reset(&mut self) -> Result<(), Error> {
        self.settings = Self::default_settings();
        let path = self.path.clone();
        self.save()
            .or_else(|e| async move {
                log::error!(
                    "{}",
                    e.display_chain_with_msg("Unable to save default settings")
                );
                log::info!("Will attempt to remove settings file");
                fs::remove_file(&path)
                    .map_err(|e| Error::DeleteError(path.display().to_string(), e))
                    .await
            })
            .await
    }

    pub fn to_settings(&self) -> Settings {
        self.settings.clone()
    }

    /// Modifies `Settings::default()` somewhat, e.g. depending on whether a beta version
    /// is being run or not.
    fn default_settings() -> Settings {
        let mut settings = Settings::default();

        if crate::version::is_beta_version() {
            settings.show_beta_releases = true;
        }
        settings
    }

    /// Edit the settings in a closure, and write the changes, if any, to disk.
    ///
    /// On success, the function returns a boolean indicating whether any settings were changed.
    /// If the settings could not be written to disk, all changes are rolled back, and an error is
    /// returned.
    pub async fn update(
        &mut self,
        update_fn: impl FnOnce(&mut Settings),
    ) -> Result<MadeChanges, Error> {
        let mut new_settings = self.settings.clone();

        update_fn(&mut new_settings);

        if self.settings == new_settings {
            return Ok(false);
        }

        Self::save_inner(&self.path, &new_settings).await?;
        self.settings = new_settings;
        Ok(true)
    }

    /// Return a compact summary of important settings
    pub fn summary(&self) -> SettingsSummary<'_> {
        SettingsSummary {
            settings: &self.settings,
        }
    }
}

impl Deref for SettingsPersister {
    type Target = Settings;

    fn deref(&self) -> &Self::Target {
        &self.settings
    }
}

/// A compact summary of important settings
pub struct SettingsSummary<'a> {
    settings: &'a Settings,
}

impl<'a> Display for SettingsSummary<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        let bool_to_label = |state| {
            if state {
                "on"
            } else {
                "off"
            }
        };

        let relay_settings = self.settings.get_relay_settings();

        write!(f, "ovpn mssfix: ")?;
        Self::fmt_option(f, self.settings.tunnel_options.openvpn.mssfix)?;
        write!(f, ", wg mtu: ")?;
        Self::fmt_option(f, self.settings.tunnel_options.wireguard.mtu)?;

        if let RelaySettings::Normal(RelayConstraints {
            wireguard_constraints: WireguardConstraints { ip_version, .. },
            ..
        }) = relay_settings
        {
            write!(f, ", wg ip: {ip_version}")?;
        }

        let multihop = matches!(
            relay_settings,
            RelaySettings::Normal(RelayConstraints {
                wireguard_constraints: WireguardConstraints {
                    use_multihop: true,
                    ..
                },
                ..
            })
        );

        write!(
            f,
            ", multihop: {}, ipv6 (tun): {}, lan: {}, pq: {}, obfs: {}",
            bool_to_label(multihop),
            bool_to_label(self.settings.tunnel_options.generic.enable_ipv6),
            bool_to_label(self.settings.allow_lan),
            self.settings.tunnel_options.wireguard.quantum_resistant,
            self.settings.obfuscation_settings.selected_obfuscation,
        )?;

        // Print DNS options

        match self.settings.tunnel_options.dns_options.state {
            DnsState::Default => {
                let mut content = vec!["default"];

                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_ads
                {
                    content.push("ads");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_trackers
                {
                    content.push("trackers");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_malware
                {
                    content.push("malware");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_adult_content
                {
                    content.push("adult");
                }
                if self
                    .settings
                    .tunnel_options
                    .dns_options
                    .default_options
                    .block_gambling
                {
                    content.push("gambling");
                }

                write!(f, ", dns: {}", content.join(" "))?;
            }
            DnsState::Custom => {
                // NOTE: Technically inaccurate, as the gateway IP is a local IP but isn't treated as one.
                let contains_local = self
                    .settings
                    .tunnel_options
                    .dns_options
                    .custom_options
                    .addresses
                    .iter()
                    .any(is_local_address);
                let contains_public = self
                    .settings
                    .tunnel_options
                    .dns_options
                    .custom_options
                    .addresses
                    .iter()
                    .any(|addr| !is_local_address(addr));

                match (contains_public, contains_local) {
                    (true, true) => f.write_str(", dns: custom, public, local")?,
                    (true, false) => f.write_str(", dns: custom, public")?,
                    (false, false) => f.write_str(", dns: custom, no addrs")?,
                    (false, true) => f.write_str(", dns: custom, local")?,
                }
            }
        }
        Ok(())
    }
}

impl<'a> SettingsSummary<'a> {
    fn fmt_option<T: Display>(f: &mut fmt::Formatter<'_>, val: Option<T>) -> fmt::Result {
        if let Some(inner) = &val {
            inner.fmt(f)
        } else {
            f.write_str("unset")
        }
    }
}

#[cfg(test)]
mod test {
    use super::SettingsPersister;
    use mullvad_types::settings::SettingsVersion;
    use serde_json;

    #[test]
    #[should_panic]
    fn test_deserialization_failure_version_too_small() {
        let _version: SettingsVersion = serde_json::from_str("1").expect("Version too small");
    }

    #[test]
    #[should_panic]
    fn test_deserialization_failure_version_too_big() {
        let _version: SettingsVersion = serde_json::from_str("1000").expect("Version too big");
    }

    #[test]
    fn test_deserialization_success() {
        let _version: SettingsVersion =
            serde_json::from_str("2").expect("Failed to deserialize valid version");
    }

    #[test]
    fn test_serialization_success() {
        let version = SettingsVersion::V2;
        let s = serde_json::to_string(&version).expect("Failed to serialize");
        assert_eq!(s, "2");
    }

    #[test]
    fn test_deserialization() {
        let settings = br#"{
              "account_token": "0000000000000000",
              "relay_settings": {
                "normal": {
                  "location": {
                    "only": {
                      "country": "gb"
                    }
                  },
                  "tunnel_protocol": {
                    "only": "wireguard"
                  },
                  "wireguard_constraints": {
                    "port": "any"
                  },
                  "openvpn_constraints": {
                    "port": "any",
                    "protocol": "any"
                  }
                }
              },
              "bridge_settings": {
                "normal": {
                  "location": "any"
                }
              },
              "bridge_state": "auto",
              "allow_lan": true,
              "block_when_disconnected": false,
              "auto_connect": true,
              "tunnel_options": {
                "openvpn": {
                  "mssfix": null
                },
                "wireguard": {
                  "mtu": null,
                  "rotation_interval": null
                },
                "generic": {
                  "enable_ipv6": true
                }
              },
              "settings_version": 5,
              "show_beta_releases": false
        }"#;

        let _ = SettingsPersister::load_from_bytes(settings).unwrap();
    }
}
