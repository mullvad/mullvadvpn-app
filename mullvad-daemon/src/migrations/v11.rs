use super::{Error, Result};
use mullvad_types::settings::SettingsVersion;

/// The migration handles:
/// - Renaming of block_when_disconnected option to lockdown_mode.
/// - API access method names must now be unique and duplicates will be renamed.
pub fn migrate(settings: &mut serde_json::Value) -> Result<()> {
    if !(version(settings) == Some(SettingsVersion::V11)) {
        return Ok(());
    }

    log::info!("Migrating settings format to v12");

    migrate_block_when_disconnected(settings)?;
    migrate_duplicated_api_access_method_names(settings)?;

    settings["settings_version"] = serde_json::json!(SettingsVersion::V12);

    Ok(())
}

fn version(settings: &serde_json::Value) -> Option<SettingsVersion> {
    settings
        .get("settings_version")
        .and_then(|version| serde_json::from_value(version.clone()).ok())
}

fn migrate_block_when_disconnected(settings: &mut serde_json::Value) -> Result<()> {
    let key_name_before = "block_when_disconnected";
    let key_name_after = "lockdown_mode";

    let settings_map = settings
        .as_object_mut()
        .ok_or(Error::InvalidSettingsContent)?;

    // Get the old key's value and insert the new key with that value
    let value = settings_map
        .get(key_name_before)
        .ok_or(Error::InvalidSettingsContent)?;
    settings_map.insert(key_name_after.to_string(), value.clone());

    // Remove the old key
    settings_map.remove(key_name_before);

    Ok(())
}

fn get_is_access_method_name_duplicate(
    used_access_method_names: &[String],
    access_method_name: &String,
) -> bool {
    used_access_method_names
        .iter()
        .filter(|used_name| *used_name == access_method_name)
        .count()
        > 0
}

fn generate_unique_access_method_name(
    used_access_method_names: &[String],
    access_method_name: &String,
    suffix: usize,
) -> String {
    // Create a new name by appending the suffix to the old name
    let unique_access_method_name = format!("{access_method_name}_{suffix}",);

    // Check if the new name is unique
    let is_access_method_name_duplicate =
        get_is_access_method_name_duplicate(used_access_method_names, &unique_access_method_name);
    if is_access_method_name_duplicate {
        // If the generated name is also a duplicate, increment the suffix and try again
        generate_unique_access_method_name(used_access_method_names, access_method_name, suffix + 1)
    } else {
        unique_access_method_name
    }
}

fn migrate_duplicated_api_access_method_names(settings: &mut serde_json::Value) -> Result<()> {
    let settings_map = settings
        .as_object_mut()
        .ok_or(Error::InvalidSettingsContent)?;

    if let Some(custom_api_access_methods) = settings_map
        .get_mut("api_access_methods")
        .and_then(serde_json::Value::as_object_mut)
        .and_then(|api_access_method| api_access_method.get_mut("custom"))
        .and_then(|custom_api_access_methods| custom_api_access_methods.as_array_mut())
    {
        // Vector to hold all API access method's names in use, it is used to figure
        // if an access method is unique or if it should be renamed
        let mut used_access_method_names: Vec<String> = Vec::new();
        for access_method in custom_api_access_methods.iter_mut() {
            if let Some(access_method_name) = access_method.get("name") {
                // Decode name from JSON
                let access_method_name_string: String =
                    serde_json::from_value(access_method_name.clone()).unwrap();

                let is_access_method_name_duplicate = get_is_access_method_name_duplicate(
                    &used_access_method_names,
                    &access_method_name_string,
                );
                // Check if access method name is a duplicate,
                // and rename it if
                if is_access_method_name_duplicate {
                    // Generate a new name for the access method
                    let unique_access_method_name = generate_unique_access_method_name(
                        &used_access_method_names,
                        &access_method_name_string,
                        1,
                    );

                    // Encode new name as JSON
                    let unique_access_method_name_json =
                        serde_json::to_value(unique_access_method_name.clone()).unwrap();

                    // Update the access method's name in settings by first removing the name key
                    // and then inserting the new value for the name key
                    let access_method_map = access_method.as_object_mut().unwrap();
                    access_method_map.remove("name");
                    access_method_map.insert("name".to_string(), unique_access_method_name_json);

                    // Add the access method's new name to the vector so we
                    // can handle if another access method would happen to use
                    // the new generated name.
                    used_access_method_names.push(unique_access_method_name);
                } else {
                    // If the access method's current name was unique, we add the
                    // name to the used names vector so we can handle if another
                    // access method also uses that name.
                    used_access_method_names.push(access_method_name_string);
                }
            }
        }
    }

    Ok(())
}

#[cfg(test)]
mod test {
    use serde_json::json;

    use crate::migrations::v11::migrate_block_when_disconnected;
    use crate::migrations::v11::migrate_duplicated_api_access_method_names;

    /// "block_when_disconnected" is renamed to "lockdown_mode"
    #[test]
    fn test_v11_to_v12_migration_block_when_disconnected_disabled() {
        let mut old_settings = json!({
            "block_when_disconnected": false,
        });
        migrate_block_when_disconnected(&mut old_settings).unwrap();
        let new_settings: serde_json::Value = json!({
            "lockdown_mode": false,
        });
        assert_eq!(&old_settings, &new_settings);
    }

    #[test]
    fn test_v11_to_v12_migration_block_when_disconnected_enabled() {
        let mut old_settings = json!({
            "block_when_disconnected": true,
        });
        migrate_block_when_disconnected(&mut old_settings).unwrap();
        let new_settings: serde_json::Value = json!({
            "lockdown_mode": true,
        });
        assert_eq!(&old_settings, &new_settings);
    }

    // custom access method's names are renamed if they are not unique
    #[test]
    fn test_v11_to_v12_migration_access_method_name_duplicates() {
        let mut old_settings = json!({
            "api_access_methods": {
              "custom": [
                  {
                  "id": "90d35296-3823-4805-8926-720fff53c752",
                  "name": "test_3",
                  "enabled": true,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:80",
                        "password": "",
                        "cipher": "aes-128-cfb"
                      }
                    }
                  }
                },
                {
                  "id": "d879f6e9-c052-4452-8e53-088183d01c0a",
                  "name": "test_2",
                  "enabled": true,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:443",
                        "password": "secret",
                        "cipher": "aes-256-cfb"
                      }
                    }
                  }
                },
                {
                  "id": "7b49cef8-5a9f-4bbc-9bee-00841edc98e9",
                  "name": "test",
                  "enabled": true,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:443",
                        "password": "",
                        "cipher": "aes-256-gcm"
                      }
                    }
                  }
                },
                 {
                  "id": "0bafc4ed-cd4f-4368-b067-74527f42451b",
                  "name": "test_1",
                  "enabled": true,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:80",
                        "password": "",
                        "cipher": "aes-128-gcm"
                      }
                    }
                  }
                },
                {
                  "id": "09d032bc-7e3a-4d85-a63f-528b6c4b890e",
                  "name": "test_2",
                  "enabled": false,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:443",
                        "password": "",
                        "cipher": "aes-128-cfb"
                      }
                    }
                  }
                },
                {
                  "id": "4471b3f5-a87e-4355-aea8-72c4c4936479",
                  "name": "test_1",
                  "enabled": false,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:80",
                        "password": "secret",
                        "cipher": "aes-256-gcm"
                      }
                    }
                  }
                },
                {
                  "id": "6f8db7c3-2258-46c0-8b7d-2016dd9e5739",
                  "name": "test",
                  "enabled": false,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:443",
                        "password": "",
                        "cipher": "aes-128-cfb"
                      }
                    }
                  }
                },
                                {
                  "id": "6f8db7c3-2258-46c0-8b7d-2016dd9e5739",
                  "name": "other_name",
                  "enabled": false,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:9090",
                        "password": "",
                        "cipher": "aes-256-cfb"
                      }
                    }
                  }
                },
                {
                  "id": "ffdf9900-e843-4298-9478-a9dfbaa63b17",
                  "name": "test",
                  "enabled": true,
                  "access_method": {
                    "custom": {
                      "shadowsocks": {
                        "endpoint": "127.0.0.1:8080",
                        "password": "",
                        "cipher": "aes-128-gcm"
                      }
                    }
                  }
                }
              ]
            }
        });
        migrate_duplicated_api_access_method_names(&mut old_settings).unwrap();
        insta::assert_snapshot!(serde_json::to_string_pretty(&old_settings).unwrap());
    }
}
