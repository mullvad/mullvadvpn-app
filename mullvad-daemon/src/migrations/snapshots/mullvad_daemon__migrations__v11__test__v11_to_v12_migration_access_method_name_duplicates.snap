---
source: mullvad-daemon/src/migrations/v11.rs
expression: "serde_json::to_string_pretty(&old_settings).unwrap()"
---
{
  "api_access_methods": {
    "custom": [
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-128-cfb",
              "endpoint": "127.0.0.1:80",
              "password": ""
            }
          }
        },
        "enabled": true,
        "id": "90d35296-3823-4805-8926-720fff53c752",
        "name": "test_3"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-256-cfb",
              "endpoint": "127.0.0.1:443",
              "password": "secret"
            }
          }
        },
        "enabled": true,
        "id": "d879f6e9-c052-4452-8e53-088183d01c0a",
        "name": "test_2"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-256-gcm",
              "endpoint": "127.0.0.1:443",
              "password": ""
            }
          }
        },
        "enabled": true,
        "id": "7b49cef8-5a9f-4bbc-9bee-00841edc98e9",
        "name": "test"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-128-gcm",
              "endpoint": "127.0.0.1:80",
              "password": ""
            }
          }
        },
        "enabled": true,
        "id": "0bafc4ed-cd4f-4368-b067-74527f42451b",
        "name": "test_1"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-128-cfb",
              "endpoint": "127.0.0.1:443",
              "password": ""
            }
          }
        },
        "enabled": false,
        "id": "09d032bc-7e3a-4d85-a63f-528b6c4b890e",
        "name": "test_2_1"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-256-gcm",
              "endpoint": "127.0.0.1:80",
              "password": "secret"
            }
          }
        },
        "enabled": false,
        "id": "4471b3f5-a87e-4355-aea8-72c4c4936479",
        "name": "test_1_1"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-128-cfb",
              "endpoint": "127.0.0.1:443",
              "password": ""
            }
          }
        },
        "enabled": false,
        "id": "6f8db7c3-2258-46c0-8b7d-2016dd9e5739",
        "name": "test_4"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-256-cfb",
              "endpoint": "127.0.0.1:9090",
              "password": ""
            }
          }
        },
        "enabled": false,
        "id": "6f8db7c3-2258-46c0-8b7d-2016dd9e5739",
        "name": "other_name"
      },
      {
        "access_method": {
          "custom": {
            "shadowsocks": {
              "cipher": "aes-128-gcm",
              "endpoint": "127.0.0.1:8080",
              "password": ""
            }
          }
        },
        "enabled": true,
        "id": "ffdf9900-e843-4298-9478-a9dfbaa63b17",
        "name": "test_5"
      }
    ]
  }
}
